components/webgpu/lib.rs

```diff
@@ -0,0 +1,128 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#[macro_use]
+extern crate serde;
+#[macro_use]
+pub extern crate wgpu_native as wgpu;
+
+use ipc_channel::ipc::{self, IpcReceiver, IpcSender};
+use malloc_size_of::{MallocSizeOf, MallocSizeOfOps};
+use servo_config::pref;
+use wgpu::adapter_get_info;
+use wgpu::TypedId;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WebGPUResponse {
+    RequestAdapter(String, WebGPUAdapter),
+    RequestDevice,
+}
+
+pub type WebGPUResponseResult = Result<WebGPUResponse, String>;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum WebGPURequest {
+    RequestAdapter(IpcSender<WebGPUResponseResult>, wgpu::RequestAdapterOptions),
+    RequestDevice,
+    Exit(IpcSender<()>),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct WebGPU(pub IpcSender<WebGPURequest>);
+
+impl WebGPU {
+    pub fn new() -> Option<Self> {
+        if !pref!(dom.webgpu.enabled) {
+            return None;
+        }
+        let (sender, receiver) = ipc::channel().unwrap();
+
+        std::thread::Builder::new()
+            .name("WGPU".to_owned())
+            .spawn(move || {
+                WGPU::new(receiver).run();
+            })
+            .expect("Thread spawning failed");
```

Comment:
Ok last comment: since this will run in the "main process" we try to not panic there as it would bring down the whole browser. So could you please change any `expect` or `unwrap` in this file into a `warn!`?

Reply:
Reploced thoose in 873a55bd066de6e984c8b6c5f11a542d5c18b61f